dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/job/NotificationSender.java

```diff
@@ -35,19 +35,83 @@
 import org.hisp.dhis.notification.logging.NotificationTriggerEvent;
 import org.hisp.dhis.notification.logging.NotificationValidationResult;
 import org.hisp.dhis.program.Enrollment;
+import org.hisp.dhis.program.Event;
+import org.hisp.dhis.program.notification.ProgramNotificationInstance;
+import org.hisp.dhis.program.notification.ProgramNotificationInstanceService;
+import org.hisp.dhis.program.notification.ProgramNotificationService;
 import org.hisp.dhis.program.notification.ProgramNotificationTemplate;
 import org.hisp.dhis.program.notification.ProgramNotificationTemplateService;
+import org.hisp.dhis.program.notification.template.snapshot.NotificationTemplateService;
 import org.hisp.dhis.programrule.api.NotificationEffect;
 import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
 
+/** Sends or schedule notification as a result of rule-engine evaluation. */
 @Slf4j
 @RequiredArgsConstructor
 @Component
-class NotificationHelper {
+public class NotificationSender {
+  public static final String LOG_MESSAGE = "Notification with id:%s has been scheduled";
+  private final ProgramNotificationInstanceService programNotificationInstanceService;
+  private final ProgramNotificationService programNotificationService;
+  private final NotificationTemplateService notificationTemplateService;
   private final ProgramNotificationTemplateService programNotificationTemplateService;
   private final NotificationLoggingService notificationLoggingService;
 
-  public void createLogEntry(ProgramNotificationTemplate template, Enrollment enrollment) {
+  @Transactional
+  public void send(NotificationEffect notificationEffect, Enrollment enrollment) {
+    ProgramNotificationTemplate template = getNotificationTemplate(notificationEffect);
+
+    NotificationValidationResult result = validate(template, enrollment);
+
+    if (result.isValid()) {
+      if (notificationEffect.date() != null) {
+        ProgramNotificationInstance notificationInstance =
```

Comment:
A thought for another day: why do we have to go through 4 lines to "send" a notification. The `programNotificationInstanceService` could provide a nicer API like `programNotificationInstanceService.sendOn(template, enrollment, date)` and `programNotificationInstanceService.sendNow(template, enrollment)`. (The current `sendProgramRuleTriggeredNotifications` is a peculiar name, don't know if it matters for the method to know what the trigger was.)

What is `createLogEntry(template, enrollment);` used for? Could this also be a responsibility of the `programNotificationInstanceService`?

The debug log below could be part of the service if needed.

Reply:
I would leave this and the log considerations to after we have a proper send/schedule async notifications service.
Platform team is looking into that